In [1]:
%cd ../..

/home/saplab/thaiminhpv/EFISS/AI-model-query-service


In [2]:
import json
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os
import sys
from tqdm.auto import tqdm
tqdm.pandas()

In [3]:
# mongoexport --uri mongodb+srv://efiss:*@efiss.wj88voa.mongodb.net/efiss --collection products --type json --out output-shopee.json --query '{"url": {"$regex": "https://shopee.vn"}}'

db_file = 'data/shopee/output-shopee.json'

data = []
with open(db_file, 'r') as f:
    for line in tqdm(f):
        data.append(json.loads(line))

0it [00:00, ?it/s]

In [4]:
bdf = pd.DataFrame(data)
bdf.dropna(subset=['originalImages'], inplace=True)
df = bdf
# df = bdf.iloc[:1000]
df

_id  \
0       {'$oid': '6460ab0252e365505c04bdd6'}   
1       {'$oid': '6460aaee52e365505c04bd06'}   
2       {'$oid': '6460aafe52e365505c04bda5'}   
3       {'$oid': '6460aaf952e365505c04bd6f'}   
4       {'$oid': '6460ab0552e365505c04be03'}   
...                                      ...   
110576  {'$oid': '6462d711614eb2054537390d'}   
110577  {'$oid': '6462d717614eb2054537394f'}   
110578  {'$oid': '6462d71b614eb20545373976'}   
110579  {'$oid': '6462d715614eb2054537393d'}   
110580  {'$oid': '6462d71a614eb2054537396d'}   

                                                    title  \
0                     28801 ÁO THUN LEN CỘC TAY KHOÉT VAI   
1                            Váy lụa tay bồng cutout lưng   
2                             Set váy yếm kèm áo tay bồng   
3                                       Chân váy thô kẻ ô   
4                           891 ÁO THUN TRƠN NHIỀU MÀU   
...                                                   ...   
110576  Áo sơ mi dài tay cổ sen 4LOVA cho bé gái kiểu ...   
110577  Quần thun cotton dài bo gấu 4LOVA cho bé trai ...   
110578  Scrunchies dây buộc tóc 4LOVA hình hoa cúc nhí...   
110579  Áo thun dài tay unisex 4LOVA nhiều họa tiết ng...   
110580  Áo thun cộc tay hoạt hình cho bé trai 4LOVA ch...   

                                                    url   price  \
0       https://shopee.vn/product/134147548/23139530788  240000   
1       https://shopee.vn/product/238591132/23240639596  480000   
2       https://shopee.vn/product/238591132/23136622940  580000   
3       https://shopee.vn/product/238591132/20877812095  294000   
4       https://shopee.vn/product/134147548/23621140991  185000   
...                                                 ...     ...   
110576   https://shopee.vn/product/244512865/5851475323   75000   
110577  https://shopee.vn/product/244512865/19167117158   70000   
110578   https://shopee.vn/product/244512865/4246255828    8000   
110579   https://shopee.vn/product/244512865/5957319697   79000   
110580  https://shopee.vn/product/244512865/17462534948   59000   

                                              description  \
0       cho len basic must  have item nàng nào cũng nê...   
1       22.De chào bạn! 🤗\nĐể biết thêm thông tin chi ...   
2       22.De chào bạn! 🤗\nĐể biết thêm thông tin chi ...   
3       22.De chào bạn! 🤗\nĐể biết thêm thông tin chi ...   
4       Must have item nàng nào cũng nên có trong tủ đ...   
...                                                   ...   
110576  - 4LOVA là shop uy tín chuyên cung cấp những s...   
110577  - 4LOVA là shop uy tín chuyên cung cấp những s...   
110578  - 4LOVA là shop uy tín chuyên cung cấp những s...   
110579  - 4LOVA là shop uy tín chuyên cung cấp những s...   
110580  - 4LOVA là shop uy tín chuyên cung cấp những s...   

                                                   images  \
0       [https://storage.googleapis.com/efiss/data/pro...   
1       [https://storage.googleapis.com/efiss/data/pro...   
2       [https://storage.googleapis.com/efiss/data/pro...   
3       [https://storage.googleapis.com/efiss/data/pro...   
4       [https://storage.googleapis.com/efiss/data/pro...   
...                                                   ...   
110576  [https://storage.googleapis.com/efiss/data/pro...   
110577  [https://storage.googleapis.com/efiss/data/pro...   
110578  [https://storage.googleapis.com/efiss/data/pro...   
110579  [https://storage.googleapis.com/efiss/data/pro...   
110580  [https://storage.googleapis.com/efiss/data/pro...   

                         group  __v  \
0              shopee-thedelia    0   
1       shopee-22decembrestore    0   
2       shopee-22decembrestore    0   
3       shopee-22decembrestore    0   
4              shopee-thedelia    0   
...                        ...  ...   
110576            shopee-4lova    0   
110577            shopee-4lova    0   
110578            shopee-4lova    0   
110579            shopee-4lova    0   
110580            shopee-4lova 

In [5]:
df = df[['title', 'images', 'categories']]
df['categories'] = df['categories'].apply(lambda x: " > ".join(x))
df

/tmp/ipykernel_1287348/3602483259.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['categories'] = df['categories'].apply(lambda x: " > ".join(x))


title  \
0                     28801 ÁO THUN LEN CỘC TAY KHOÉT VAI   
1                            Váy lụa tay bồng cutout lưng   
2                             Set váy yếm kèm áo tay bồng   
3                                       Chân váy thô kẻ ô   
4                           891 ÁO THUN TRƠN NHIỀU MÀU   
...                                                   ...   
110576  Áo sơ mi dài tay cổ sen 4LOVA cho bé gái kiểu ...   
110577  Quần thun cotton dài bo gấu 4LOVA cho bé trai ...   
110578  Scrunchies dây buộc tóc 4LOVA hình hoa cúc nhí...   
110579  Áo thun dài tay unisex 4LOVA nhiều họa tiết ng...   
110580  Áo thun cộc tay hoạt hình cho bé trai 4LOVA ch...   

                                                   images  \
0       [https://storage.googleapis.com/efiss/data/pro...   
1       [https://storage.googleapis.com/efiss/data/pro...   
2       [https://storage.googleapis.com/efiss/data/pro...   
3       [https://storage.googleapis.com/efiss/data/pro...   
4       [https://storage.googleapis.com/efiss/data/pro...   
...                                                   ...   
110576  [https://storage.googleapis.com/efiss/data/pro...   
110577  [https://storage.googleapis.com/efiss/data/pro...   
110578  [https://storage.googleapis.com/efiss/data/pro...   
110579  [https://storage.googleapis.com/efiss/data/pro...   
110580  [https://storage.googleapis.com/efiss/data/pro...   

                                               categories  
0                               Thời Trang Nữ > Áo > Khác  
1                                     Thời Trang Nữ > Đầm  
2                 Thời Trang Nữ > Đồ liền thân > Quần yếm  
3                                     Thời Trang Nữ > Váy  
4                            Thời Trang Nữ > Áo > Áo thun  
...                                                   ...  
110576  Thời trang trẻ em & trẻ sơ sinh > Quần áo bé g...  
110577  Thời trang trẻ em & trẻ sơ sinh > Quần áo bé t...  
110578  Phụ Kiện Thời Trang > Phụ kiện tóc > Đồ buộc t...  
110579  Thời trang trẻ em & trẻ sơ sinh > Quần áo trẻ ...  
110580  Thời trang trẻ em & trẻ sơ sinh > Quần áo bé t...  

[110434 rows x 3 columns]

In [6]:
df['categories'].value_counts()

Thời Trang Nữ > Đầm                                                       10281
Thời Trang Nữ > Áo > Áo thun                                               4906
Thời Trang Nữ > Áo > Áo sơ mi                                              4847
Thời Trang Nam > Áo > Áo sơ mi                                             4522
Phụ Kiện Thời Trang > Kính mắt > Gọng kính                                 4304
                                                                          ...  
Nhà cửa & Đời sống > Sắp xếp nhà cửa > Kệ giày, hộp giày                      1
Thực phẩm và đồ uống > Nguyên liệu nấu ăn > Gia vị & Hương liệu > Giấm        1
Thực phẩm và đồ uống > Đồ làm bánh > Bột nở và muối nở                        1
Sắc Đẹp > Chăm sóc nam giới > Sữa tắm & chăm sóc cơ thể                       1
Mẹ & Bé > Đồ dùng du lịch cho bé > Xe đẩy                                     1
Name: categories, Length: 702, dtype: int64

In [7]:
temp = df['categories'].value_counts()
temp = temp[temp > 10]
temp

Thời Trang Nữ > Đầm                                            10281
Thời Trang Nữ > Áo > Áo thun                                    4906
Thời Trang Nữ > Áo > Áo sơ mi                                   4847
Thời Trang Nam > Áo > Áo sơ mi                                  4522
Phụ Kiện Thời Trang > Kính mắt > Gọng kính                      4304
                                                               ...  
Mẹ & Bé > Chăm sóc sức khỏe bé > Chăm sóc răng miệng cho bé       12
Túi Ví Nữ > Phụ kiện túi > Túi đa ngăn tiện ích                   12
Nhà cửa & Đời sống > Bộ đồ bàn ăn > Cốc, ly, tách uống nước       12
Nhà cửa & Đời sống > Trang trí nhà cửa > Thảm trải sàn            11
Thể Thao & Dã Ngoại > Giày Thể Thao > Giày Tennis                 11
Name: categories, Length: 333, dtype: int64

In [ ]:
# temp.to_csv('data/shopee/cls.csv', index=True)

In [23]:
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(temp)

                                                    categories
Thời Trang Nữ > Đầm                                      10281
Thời Trang Nữ > Áo > Áo thun                              4906
Thời Trang Nữ > Áo > Áo sơ mi                             4847
Thời Trang Nam > Áo > Áo sơ mi                            4522
Phụ Kiện Thời Trang > Kính mắt > Gọng kính                4304
Thời Trang Nam > Áo > Áo thun                             4176
Thời Trang Nữ > Váy                                       3725
Thời Trang Nam > Áo > Áo polo                             3359
Phụ Kiện Thời Trang > Kính mắt > Kính mát                 2886
Thời Trang Nữ > Đồ lót > Quần lót                         2568
Thời Trang Nữ > Đồ lót > Áo ngực                          2532
Thời trang trẻ em & trẻ sơ sinh > Quần áo trẻ e...        2400
Thời trang trẻ em & trẻ sơ sinh > Quần áo bé gá...        2293
Thời Trang Nữ > Quần > Quần dài                           2221
Thời Trang Nam > Quần đùi                              

... After manually filter all fashion-unrelated labels

In [8]:
temp = pd.read_csv('data/shopee/cls.csv', index_col=0)
temp

categories
Thời Trang Nữ > Đầm                                      10281
Thời Trang Nữ > Áo > Áo thun                              4906
Thời Trang Nữ > Áo > Áo sơ mi                             4847
Thời Trang Nam > Áo > Áo sơ mi                            4522
Phụ Kiện Thời Trang > Kính mắt > Gọng kính                4304
...                                                        ...
Thời Trang Nam > Áo khoác > Áo khoác vest                   13
Túi Ví Nữ > Phụ kiện túi > Phụ kiện khác                    13
Thể Thao & Dã Ngoại > Thời Trang Thể Thao & Dã ...          13
Thể Thao & Dã Ngoại > Giày Thể Thao > Giày Tập ...          13
Thể Thao & Dã Ngoại > Giày Thể Thao > Giày Tennis           11

[272 rows x 1 columns]

### We have 272 class left, we will use them

How many ``images = []`` is missing?

In [9]:
sum(df['images'].apply(lambda x: len(x)) == 0) / len(df)

0.0

In [10]:
sum(bdf['originalImages'].apply(lambda x: len(x)) == 0) / len(df)

0.0

### Generate dataset for training

In [11]:
bdf.shape

(110434, 10)

In [12]:
temp.index

Index(['Thời Trang Nữ > Đầm', 'Thời Trang Nữ > Áo > Áo thun',
       'Thời Trang Nữ > Áo > Áo sơ mi', 'Thời Trang Nam > Áo > Áo sơ mi',
       'Phụ Kiện Thời Trang > Kính mắt > Gọng kính',
       'Thời Trang Nam > Áo > Áo thun', 'Thời Trang Nữ > Váy',
       'Thời Trang Nam > Áo > Áo polo',
       'Phụ Kiện Thời Trang > Kính mắt > Kính mát',
       'Thời Trang Nữ > Đồ lót > Quần lót',
       ...
       'Thời Trang Nữ > Đồ Bầu > Khác',
       'Phụ Kiện Thời Trang > Phụ kiện tóc > Tóc giả & Tóc nối',
       'Thời Trang Nam > Áo khoác > Khác',
       'Chăm Sóc Thú Cưng > Quần áo & phụ kiện > Kính mắt',
       'Túi Ví Nữ > Ví > Ví đựng điện thoại & chìa khóa',
       'Thời Trang Nam > Áo khoác > Áo khoác vest',
       'Túi Ví Nữ > Phụ kiện túi > Phụ kiện khác',
       'Thể Thao & Dã Ngoại > Thời Trang Thể Thao & Dã Ngoại > Đồ Bơi > Đồ Bơi Khác',
       'Thể Thao & Dã Ngoại > Giày Thể Thao > Giày Tập Luyện',
       'Thể Thao & Dã Ngoại > Giày Thể Thao > Giày Tennis'],
      dtype='object', 

In [13]:
adf = df[df['categories'].isin(temp.index)]
adf

title  \
0                     28801 ÁO THUN LEN CỘC TAY KHOÉT VAI   
1                            Váy lụa tay bồng cutout lưng   
2                             Set váy yếm kèm áo tay bồng   
3                                       Chân váy thô kẻ ô   
4                           891 ÁO THUN TRƠN NHIỀU MÀU   
...                                                   ...   
110576  Áo sơ mi dài tay cổ sen 4LOVA cho bé gái kiểu ...   
110577  Quần thun cotton dài bo gấu 4LOVA cho bé trai ...   
110578  Scrunchies dây buộc tóc 4LOVA hình hoa cúc nhí...   
110579  Áo thun dài tay unisex 4LOVA nhiều họa tiết ng...   
110580  Áo thun cộc tay hoạt hình cho bé trai 4LOVA ch...   

                                                   images  \
0       [https://storage.googleapis.com/efiss/data/pro...   
1       [https://storage.googleapis.com/efiss/data/pro...   
2       [https://storage.googleapis.com/efiss/data/pro...   
3       [https://storage.googleapis.com/efiss/data/pro...   
4       [https://storage.googleapis.com/efiss/data/pro...   
...                                                   ...   
110576  [https://storage.googleapis.com/efiss/data/pro...   
110577  [https://storage.googleapis.com/efiss/data/pro...   
110578  [https://storage.googleapis.com/efiss/data/pro...   
110579  [https://storage.googleapis.com/efiss/data/pro...   
110580  [https://storage.googleapis.com/efiss/data/pro...   

                                               categories  
0                               Thời Trang Nữ > Áo > Khác  
1                                     Thời Trang Nữ > Đầm  
2                 Thời Trang Nữ > Đồ liền thân > Quần yếm  
3                                     Thời Trang Nữ > Váy  
4                            Thời Trang Nữ > Áo > Áo thun  
...                                                   ...  
110576  Thời trang trẻ em & trẻ sơ sinh > Quần áo bé g...  
110577  Thời trang trẻ em & trẻ sơ sinh > Quần áo bé t...  
110578  Phụ Kiện Thời Trang > Phụ kiện tóc > Đồ buộc t...  
110579  Thời trang trẻ em & trẻ sơ sinh > Quần áo trẻ ...  
110580  Thời trang trẻ em & trẻ sơ sinh > Quần áo bé t...  

[107356 rows x 3 columns]

In [14]:
adf.shape, df.shape

((107356, 3), (110434, 3))

Data reduce from 110434 to 107356

In [15]:
# convert categories to class index
class_to_idx = {c: i for i, c in enumerate(temp.index)}
class_to_idx

{'Thời Trang Nữ > Đầm': 0,
 'Thời Trang Nữ > Áo > Áo thun': 1,
 'Thời Trang Nữ > Áo > Áo sơ mi': 2,
 'Thời Trang Nam > Áo > Áo sơ mi': 3,
 'Phụ Kiện Thời Trang > Kính mắt > Gọng kính': 4,
 'Thời Trang Nam > Áo > Áo thun': 5,
 'Thời Trang Nữ > Váy': 6,
 'Thời Trang Nam > Áo > Áo polo': 7,
 'Phụ Kiện Thời Trang > Kính mắt > Kính mát': 8,
 'Thời Trang Nữ > Đồ lót > Quần lót': 9,
 'Thời Trang Nữ > Đồ lót > Áo ngực': 10,
 'Thời trang trẻ em & trẻ sơ sinh > Quần áo trẻ em > Bộ quần áo': 11,
 'Thời trang trẻ em & trẻ sơ sinh > Quần áo bé gái > Váy': 12,
 'Thời Trang Nữ > Quần > Quần dài': 13,
 'Thời Trang Nam > Quần đùi': 14,
 'Thời Trang Nữ > Bộ > Đồ lẻ': 15,
 'Thời Trang Nữ > Áo > Áo hai dây và ba lỗ': 16,
 'Thời Trang Nữ > Đồ ngủ > Pyjama': 17,
 'Thời Trang Nam > Áo khoác > Áo khoác': 18,
 'Phụ Kiện Thời Trang > Thắt lưng': 19,
 'Thời Trang Nam > Đồ lót > Quần lót': 20,
 'Thời Trang Nữ > Quần đùi > Quần đùi': 21,
 'Thời trang trẻ em & trẻ sơ sinh > Quần áo bé gái > Com lê & đồ bộ': 22,
 'T

In [16]:
# convert categories to class index
adf['class_idx'] = adf['categories'].apply(lambda x: class_to_idx[x])
adf

/tmp/ipykernel_1287348/357757967.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  adf['class_idx'] = adf['categories'].apply(lambda x: class_to_idx[x])


title  \
0                     28801 ÁO THUN LEN CỘC TAY KHOÉT VAI   
1                            Váy lụa tay bồng cutout lưng   
2                             Set váy yếm kèm áo tay bồng   
3                                       Chân váy thô kẻ ô   
4                           891 ÁO THUN TRƠN NHIỀU MÀU   
...                                                   ...   
110576  Áo sơ mi dài tay cổ sen 4LOVA cho bé gái kiểu ...   
110577  Quần thun cotton dài bo gấu 4LOVA cho bé trai ...   
110578  Scrunchies dây buộc tóc 4LOVA hình hoa cúc nhí...   
110579  Áo thun dài tay unisex 4LOVA nhiều họa tiết ng...   
110580  Áo thun cộc tay hoạt hình cho bé trai 4LOVA ch...   

                                                   images  \
0       [https://storage.googleapis.com/efiss/data/pro...   
1       [https://storage.googleapis.com/efiss/data/pro...   
2       [https://storage.googleapis.com/efiss/data/pro...   
3       [https://storage.googleapis.com/efiss/data/pro...   
4       [https://storage.googleapis.com/efiss/data/pro...   
...                                                   ...   
110576  [https://storage.googleapis.com/efiss/data/pro...   
110577  [https://storage.googleapis.com/efiss/data/pro...   
110578  [https://storage.googleapis.com/efiss/data/pro...   
110579  [https://storage.googleapis.com/efiss/data/pro...   
110580  [https://storage.googleapis.com/efiss/data/pro...   

                                               categories  class_idx  
0                               Thời Trang Nữ > Áo > Khác         24  
1                                     Thời Trang Nữ > Đầm          0  
2                 Thời Trang Nữ > Đồ liền thân > Quần yếm        112  
3                                     Thời Trang Nữ > Váy          6  
4                            Thời Trang Nữ > Áo > Áo thun          1  
...                                                   ...        ...  
110576  Thời trang trẻ em & trẻ sơ sinh > Quần áo bé g...        121  
110577  Thời trang trẻ em & trẻ sơ sinh > Quần áo bé t...         64  
110578  Phụ Kiện Thời Trang > Phụ kiện tóc > Đồ buộc t...        181  
110579  Thời trang trẻ em & trẻ sơ sinh > Quần áo trẻ ...         35  
110580  Thời trang trẻ em & trẻ sơ sinh > Quần áo bé t...         31  

[107356 rows x 4 columns]

In [17]:
adf

title  \
0                     28801 ÁO THUN LEN CỘC TAY KHOÉT VAI   
1                            Váy lụa tay bồng cutout lưng   
2                             Set váy yếm kèm áo tay bồng   
3                                       Chân váy thô kẻ ô   
4                           891 ÁO THUN TRƠN NHIỀU MÀU   
...                                                   ...   
110576  Áo sơ mi dài tay cổ sen 4LOVA cho bé gái kiểu ...   
110577  Quần thun cotton dài bo gấu 4LOVA cho bé trai ...   
110578  Scrunchies dây buộc tóc 4LOVA hình hoa cúc nhí...   
110579  Áo thun dài tay unisex 4LOVA nhiều họa tiết ng...   
110580  Áo thun cộc tay hoạt hình cho bé trai 4LOVA ch...   

                                                   images  \
0       [https://storage.googleapis.com/efiss/data/pro...   
1       [https://storage.googleapis.com/efiss/data/pro...   
2       [https://storage.googleapis.com/efiss/data/pro...   
3       [https://storage.googleapis.com/efiss/data/pro...   
4       [https://storage.googleapis.com/efiss/data/pro...   
...                                                   ...   
110576  [https://storage.googleapis.com/efiss/data/pro...   
110577  [https://storage.googleapis.com/efiss/data/pro...   
110578  [https://storage.googleapis.com/efiss/data/pro...   
110579  [https://storage.googleapis.com/efiss/data/pro...   
110580  [https://storage.googleapis.com/efiss/data/pro...   

                                               categories  class_idx  
0                               Thời Trang Nữ > Áo > Khác         24  
1                                     Thời Trang Nữ > Đầm          0  
2                 Thời Trang Nữ > Đồ liền thân > Quần yếm        112  
3                                     Thời Trang Nữ > Váy          6  
4                            Thời Trang Nữ > Áo > Áo thun          1  
...                                                   ...        ...  
110576  Thời trang trẻ em & trẻ sơ sinh > Quần áo bé g...        121  
110577  Thời trang trẻ em & trẻ sơ sinh > Quần áo bé t...         64  
110578  Phụ Kiện Thời Trang > Phụ kiện tóc > Đồ buộc t...        181  
110579  Thời trang trẻ em & trẻ sơ sinh > Quần áo trẻ ...         35  
110580  Thời trang trẻ em & trẻ sơ sinh > Quần áo bé t...         31  

[107356 rows x 4 columns]

In [18]:
pdf = adf[['images', 'class_idx']]
pdf = pdf.explode('images')
pdf

images  class_idx
0       https://storage.googleapis.com/efiss/data/prod...         24
0       https://storage.googleapis.com/efiss/data/prod...         24
0       https://storage.googleapis.com/efiss/data/prod...         24
0       https://storage.googleapis.com/efiss/data/prod...         24
1       https://storage.googleapis.com/efiss/data/prod...          0
...                                                   ...        ...
110580  https://storage.googleapis.com/efiss/data/prod...         31
110580  https://storage.googleapis.com/efiss/data/prod...         31
110580  https://storage.googleapis.com/efiss/data/prod...         31
110580  https://storage.googleapis.com/efiss/data/prod...         31
110580  https://storage.googleapis.com/efiss/data/prod...         31

[738613 rows x 2 columns]

In [19]:
pdf['images'] = "data/shopee/product_images" + pdf['images'].str.split('product_images').str[-1]
pdf['images']

0         data/shopee/product_images/shopee-thedelia/646...
0         data/shopee/product_images/shopee-thedelia/646...
0         data/shopee/product_images/shopee-thedelia/646...
0         data/shopee/product_images/shopee-thedelia/646...
1         data/shopee/product_images/shopee-22decembrest...
                                ...                        
110580    data/shopee/product_images/shopee-4lova/6462d7...
110580    data/shopee/product_images/shopee-4lova/6462d7...
110580    data/shopee/product_images/shopee-4lova/6462d7...
110580    data/shopee/product_images/shopee-4lova/6462d7...
110580    data/shopee/product_images/shopee-4lova/6462d7...
Name: images, Length: 738613, dtype: object

In [20]:
pdf.set_index('images', inplace=True)

In [21]:
pdf

class_idx
images                                                       
data/shopee/product_images/shopee-thedelia/6460...         24
data/shopee/product_images/shopee-thedelia/6460...         24
data/shopee/product_images/shopee-thedelia/6460...         24
data/shopee/product_images/shopee-thedelia/6460...         24
data/shopee/product_images/shopee-22decembresto...          0
...                                                       ...
data/shopee/product_images/shopee-4lova/6462d71...         31
data/shopee/product_images/shopee-4lova/6462d71...         31
data/shopee/product_images/shopee-4lova/6462d71...         31
data/shopee/product_images/shopee-4lova/6462d71...         31
data/shopee/product_images/shopee-4lova/6462d71...         31

[738613 rows x 1 columns]

In [22]:
pdf.reset_index(drop=False, inplace=True)
pdf

images  class_idx
0       data/shopee/product_images/shopee-thedelia/646...         24
1       data/shopee/product_images/shopee-thedelia/646...         24
2       data/shopee/product_images/shopee-thedelia/646...         24
3       data/shopee/product_images/shopee-thedelia/646...         24
4       data/shopee/product_images/shopee-22decembrest...          0
...                                                   ...        ...
738608  data/shopee/product_images/shopee-4lova/6462d7...         31
738609  data/shopee/product_images/shopee-4lova/6462d7...         31
738610  data/shopee/product_images/shopee-4lova/6462d7...         31
738611  data/shopee/product_images/shopee-4lova/6462d7...         31
738612  data/shopee/product_images/shopee-4lova/6462d7...         31

[738613 rows x 2 columns]

In [ ]:
pdf.to_csv('data/shopee/shopee-data-for-training.csv', index=False)

In [ ]:
import os
from PIL import Image


def check_img_integrity(img_path):
    try:
        with open(img_path, 'rb') as f:
            check_chars = f.read()[-2:]
        if check_chars != b'\xff\xd9':
            print(f"Not complete image: {img_path}")
            return False

        with Image.open(img_path) as img:
            img.verify()
            image_format = img.format.lower()
            if image_format == 'jpeg' and not (img_path.endswith('.jpg') or img_path.endswith('.jpeg')):
                print(f"Wrong image extension: {img_path}")
                return False

        return True
    except Exception as e:
        print(e)
        return False

In [ ]:
test = pdf.head(1000)
test

In [45]:
mask = pdf.images.progress_apply(check_img_integrity)

  0%|          | 0/738613 [00:00<?, ?it/s]

[Errno 2] No such file or directory: 'data/shopee/product_images/shopee-22decembrestore/6460aae452e365505c04bca9_4_shopee_22decembrestore.jpeg'
[Errno 2] No such file or directory: 'data/shopee/product_images/shopee-22decembrestore/6460aaf852e365505c04bd66_4_shopee_22decembrestore.jpeg'
[Errno 2] No such file or directory: 'data/shopee/product_images/shopee-thedelia/6460ab0152e365505c04bdc4_6_shopee_thedelia.jpeg'
[Errno 2] No such file or directory: 'data/shopee/product_images/shopee-thedelia/6460ab0152e365505c04bdc4_7_shopee_thedelia.jpeg'
[Errno 2] No such file or directory: 'data/shopee/product_images/shopee-thedelia/6460ab0152e365505c04bdc4_8_shopee_thedelia.jpeg'
[Errno 2] No such file or directory: 'data/shopee/product_images/shopee-rame_vn/6460c04a48de6f0ca36bea35_7_shopee_rame_vn.jpeg'
[Errno 2] No such file or directory: 'data/shopee/product_images/shopee-rame_vn/6460c04a48de6f0ca36bea35_8_shopee_rame_vn.jpeg'
[Errno 2] No such file or directory: 'data/shopee/product_images/s

In [50]:
mask

0         True
1         True
2         True
3         True
4         True
          ... 
738608    True
738609    True
738610    True
738611    True
738612    True
Name: images, Length: 738613, dtype: bool

In [46]:
print(f"Corrupted images: {(~mask).sum()}")

Corrupted images: 2573


In [47]:
print(f"Num images: {len(mask)}")

Num images: 738613


In [56]:
database = pdf[mask]

In [57]:
database.to_csv('data/shopee/shopee-data-for-training-filter-corrupted.csv', index=False)

In [58]:
database

images  class_idx
0       data/shopee/product_images/shopee-thedelia/646...         24
1       data/shopee/product_images/shopee-thedelia/646...         24
2       data/shopee/product_images/shopee-thedelia/646...         24
3       data/shopee/product_images/shopee-thedelia/646...         24
4       data/shopee/product_images/shopee-22decembrest...          0
...                                                   ...        ...
738608  data/shopee/product_images/shopee-4lova/6462d7...         31
738609  data/shopee/product_images/shopee-4lova/6462d7...         31
738610  data/shopee/product_images/shopee-4lova/6462d7...         31
738611  data/shopee/product_images/shopee-4lova/6462d7...         31
738612  data/shopee/product_images/shopee-4lova/6462d7...         31

[736040 rows x 2 columns]

In [59]:
database.to_csv('data/shopee/database.txt', sep=' ', header=False, index=False)

In [29]:
# database = pd.read_csv('data/shopee/database.txt', sep=' ', header=None, names=['images', 'class_idx'])

In [39]:
# sample 1000 images from each class, if the class has less than 1000 images,
# we sample 
def sample(df):
    if len(df) > 1000:
        return df.sample(n=30)
    elif len(df) > 300:
        return df.sample(n=10)
    elif len(df) > 100:
        return df.sample(n=5)
    elif len(df) > 5:
        return df.sample(n=2)
    else:
        return df

In [40]:
test_df = database.groupby(['class_idx']).progress_apply(sample).reset_index(drop=True)

  0%|          | 0/271 [00:00<?, ?it/s]

In [41]:
test_df

images  class_idx
0     data/shopee/product_images/shopee-afterbefore_...          0
1     data/shopee/product_images/shopee-lovito.vn/64...          0
2     data/shopee/product_images/shopee-tkfashion.mk...          0
3     data/shopee/product_images/shopee-somehow/6462...          0
4     data/shopee/product_images/shopee-sixdoofficia...          0
...                                                 ...        ...
4211  data/shopee/product_images/shopee-gymme.offici...        269
4212  data/shopee/product_images/shopee-hoangphucint...        270
4213  data/shopee/product_images/shopee-hoangphucint...        270
4214  data/shopee/product_images/shopee-thoitrangthe...        271
4215  data/shopee/product_images/shopee-thoitrangthe...        271

[4216 rows x 2 columns]

In [42]:
test_df.to_csv('data/shopee/test.txt', sep=' ', header=False, index=False)

In [43]:
!cat data/shopee/database.txt | head

data/shopee/product_images/shopee-thedelia/6460ab0252e365505c04bdd6_0_shopee_thedelia.jpeg 24
data/shopee/product_images/shopee-thedelia/6460ab0252e365505c04bdd6_1_shopee_thedelia.jpeg 24
data/shopee/product_images/shopee-thedelia/6460ab0252e365505c04bdd6_2_shopee_thedelia.jpeg 24
data/shopee/product_images/shopee-thedelia/6460ab0252e365505c04bdd6_3_shopee_thedelia.jpeg 24
data/shopee/product_images/shopee-22decembrestore/6460aaee52e365505c04bd06_0_shopee_22decembrestore.jpeg 0
data/shopee/product_images/shopee-22decembrestore/6460aaee52e365505c04bd06_1_shopee_22decembrestore.jpeg 0
data/shopee/product_images/shopee-22decembrestore/6460aaee52e365505c04bd06_2_shopee_22decembrestore.jpeg 0
data/shopee/product_images/shopee-22decembrestore/6460aafe52e365505c04bda5_0_shopee_22decembrestore.jpeg 112
data/shopee/product_images/shopee-22decembrestore/6460aafe52e365505c04bda5_1_shopee_22decembrestore.jpeg 112
data/shopee/product_images/shopee-22decembrestore/6460aafe52e365505c04bda5_2_shopee_2

In [44]:
!cat data/shopee/test.txt | head

data/shopee/product_images/shopee-afterbefore_official/6462275c614eb20545341f1e_7_shopee_afterbefore_official.jpeg 0
data/shopee/product_images/shopee-lovito.vn/64626c2f614eb20545354282_6_shopee_lovito_vn.jpeg 0
data/shopee/product_images/shopee-tkfashion.mkt/6462a2c3614eb20545362213_1_shopee_tkfashion_mkt.jpeg 0
data/shopee/product_images/shopee-somehow/64621cdb614eb20545338eb1_4_shopee_somehow.jpeg 0
data/shopee/product_images/shopee-sixdoofficial/646299fa614eb2054535fa92_3_shopee_sixdoofficial.jpeg 0
data/shopee/product_images/shopee-ivymoda/64621153614eb2054532f072_6_shopee_ivymoda.jpeg 0
data/shopee/product_images/shopee-lamer.fashion/64626d21614eb20545354e95_2_shopee_lamer_fashion.jpeg 0
data/shopee/product_images/shopee-gagobrand/646251f8614eb2054534aca3_2_shopee_gagobrand.jpeg 0
data/shopee/product_images/shopee-g.e.m.clothing/646253dc614eb2054534b569_7_shopee_g_e_m_clothing.jpeg 0
data/shopee/product_images/shopee-lovito.vn/64626a4a614eb205453527f7_5_shopee_lovito_vn.jpeg 0
ca